x-pack/plugins/task_manager/server/kibana_discovery_service/kibana_discovery_service.ts

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { ISavedObjectsRepository } from '@kbn/core/server';
+import { Logger } from '@kbn/core/server';
+import { BACKGROUND_TASK_NODE_SO_NAME } from '../saved_objects';
+import { BackgroundTaskNode } from '../saved_objects/schemas/background_task_node';
+
+interface DiscoveryServiceParams {
+  currentNode: string;
+  savedObjectsRepository: ISavedObjectsRepository;
+  logger: Logger;
+}
+
+interface DiscoveryServiceUpsertParams {
+  id: string;
+  lastSeen: string;
+}
+
+export const discoveryInterval = 1000 * 10;
+export const cleanupInterval = 1000 * 60;
+
+export class KibanaDiscoveryService {
+  private currentNode: string;
+  private discoveryStarted = false;
+  private cleanupStarted = false;
+  private savedObjectsRepository: ISavedObjectsRepository;
+  private logger: Logger;
+
+  constructor({ currentNode, savedObjectsRepository, logger }: DiscoveryServiceParams) {
+    this.savedObjectsRepository = savedObjectsRepository;
+    this.logger = logger;
+    this.currentNode = currentNode;
+  }
+
+  private async upsertCurrentNode({ id, lastSeen }: DiscoveryServiceUpsertParams) {
+    await this.savedObjectsRepository.update<BackgroundTaskNode>(
+      BACKGROUND_TASK_NODE_SO_NAME,
+      id,
+      {
+        id,
+        last_seen: lastSeen,
+      },
+      { upsert: { id, last_seen: lastSeen }, refresh: true }
+    );
+  }
+
+  private async deleteInactiveNodes() {
+    const { saved_objects: inactiveNodes } =
+      await this.savedObjectsRepository.find<BackgroundTaskNode>({
+        type: BACKGROUND_TASK_NODE_SO_NAME,
+        perPage: 10000,
+        page: 1,
+        filter: `${BACKGROUND_TASK_NODE_SO_NAME}.attributes.last_seen < now-5m`,
+      });
+
+    if (inactiveNodes.length > 0) {
+      const nodesToDelete = inactiveNodes.map((node) => ({
+        type: BACKGROUND_TASK_NODE_SO_NAME,
+        id: node.attributes.id,
+      }));
+      await this.savedObjectsRepository.bulkDelete(nodesToDelete, { force: true, refresh: true });
+      this.logger.info(
+        `Inactive Kibana nodes: ${nodesToDelete.map(
+          (node) => node.id
+        )}, have been successfully deleted`
+      );
+    }
+  }
```

Comment:
I wonder if we should make the inactive node deletion part of a background task. When we scale to 64 or even 150 Kibana nodes running background tasks, this will become inefficient and cause extra load on Elasticsearch that isn't necessary. Wdyt?

Example of background task doing cleanup work: https://github.com/elastic/kibana/tree/main/x-pack/plugins/alerting/server/invalidate_pending_api_keys